core/src/main/scala/org/http4s/Message.scala

```diff
@@ -66,7 +66,7 @@ sealed trait Message extends MessageOps { self =>
   def withBody[T](b: T)(implicit w: EntityEncoder[T]): Task[Self] = {
     w.toEntity(b).map { entity =>
       val hs = entity.length match {
-        case Some(l) => `Content-Length`(l)::w.headers.toList
+        case Some(l) => `Content-Length`(l).fold(_ => w.headers.toList, _ :: w.headers.toList)
```

Comment:
A disabled log statement is negligible, whereas a Task of a disabled log statement will never be elided.  So one has to be careful logging in a task in a hotspot.  That said, it's only going to happen in rare cases here, so I'd probably do it right and do it in a Task.